### ISSUE TITLE
Weibull.mode returns nan or a wrong value for concentration < 1, Kumaraswamy.mode is always nan

### ISSUE THREAD

**Issue by Nicholas022400701** [pytorch/pytorch#198186]
### 🐛 Describe the bug

Two `mode` properties in `torch.distributions` return wrong values. The generic `test_mode` does not catch either because it treats nan as "undefined" and samples instead, and its random Weibull example never hits the concentrations where the wrong value is finite.

**Weibull.** `Weibull.mode` evaluates `scale * ((concentration - 1) / concentration) ** (1 / concentration)` for every concentration. The closed form only holds for `concentration > 1`. Below 1 the base is negative, so `pow` returns nan for most values and a finite wrong number whenever `1 / concentration` is an integer:

```python
import torch
from torch.distributions import Weibull

Weibull(torch.tensor(2.0), torch.tensor(0.7)).mode   # tensor(nan)
Weibull(torch.tensor(2.0), torch.tensor(0.5)).mode   # tensor(2.)      (-1) ** 2 = 1
Weibull(torch.tensor(2.0), torch.tensor(0.25)).mode  # tensor(162.)    (-3) ** 4 = 81
Weibull(torch.tensor(2.0), torch.tensor(0.2)).mode   # tensor(-2048.)  (-4) ** 5 = -1024, outside the support
```

For `concentration < 1` the density is unbounded at 0, and 0 lies outside `constraints.positive`, so the mode is undefined there. `Kumaraswamy.mode` and `FisherSnedecor.mode` handle the same situation by returning nan. The patch below returns nan for `concentration < 1` and keeps the closed form for `concentration >= 1` (1 still gives 0, as today). The computation goes through a `torch.where` with a safe concentration so the nan does not leak into the gradient of `scale`. If you prefer the convention from the usual tables, 0 for `concentration <= 1`, that is a one-line change, but 0 fails the `positive` support check in `log_prob`, so `test_mode` would need an exception for it.

**Kumaraswamy.** `Kumaraswamy.mode` is nan for every parameter value, including the ordinary case where both concentrations are above 1:

```python
from torch.distributions import Kumaraswamy

Kumaraswamy(torch.tensor(2.0), torch.tensor(3.0)).mode   # tensor(nan), the density peaks at sqrt(1 / 5) = 0.4472
Kumaraswamy(torch.tensor(10.0), torch.tensor(10.0)).mode # tensor(nan), the density peaks at 0.7868
```

The mode of Kumaraswamy(a, b) is `((a - 1) / (a * b - 1)) ** (1 / a)` with `a = concentration1`. The log-space expression in the code is `log1p(-b) / b - log1p(-a * b)`, which has the wrong signs in both logarithms and uses `concentration0` as the exponent, so `log1p(-b)` is nan as soon as `b > 1`. The patch below writes `(log(a - 1) - log(a * b - 1)) / a` and keeps the existing nan rule for `a < 1` or `b < 1`. The result matches the argmax of `log_prob` on a 2,000,001 point grid for (a, b) in (2, 3), (3, 2), (1.5, 1.5), (10, 10), gives 0 for (1, 5), 1 for (5, 1) and nan for (1, 1), where the density is flat.

Fix and tests are ready (patch below): `test_weibull_mode` covers concentration 0.5, 1, 1.5 and 3, `test_kumaraswamy_mode` covers (2, 3), (3, 2), (1, 5), (5, 1), (0.5, 2) and (1, 1). Both fail on main and pass with the fix, `test_mode` and the other Weibull and Kumaraswamy tests still pass, and the perturbation check inside `test_mode` passes for the Kumaraswamy example parameters over 200 seeds. `ruff format --check` and `ruff check` with the pinned ruff 0.14.4 are clean. Would you take a PR for this?

### Versions

`Weibull.mode` and `Kumaraswamy.mode` are unchanged on main (a402343). Reproduced with the wheel below and with the files from main.

<details><summary>collect_env</summary>

```
Collecting environment information...
PyTorch version: 2.14.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2023.12.20260817 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.14 (main, Jul 15 2026, 00:00:00) [GCC 11.5.0 20240719 (Red Hat 11.5.0-5)] (64-bit runtime)
Python platform: Linux-6.18.49-x86_64-with-glibc2.34
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  2
On-line CPU(s) list:                     0,1
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Processor @ 2.50GHz
CPU family:                              6
Model:                                   85
Thread(s) per core:                      2
Core(s) per socket:                      1
Socket(s):                               1
```

</details>

### Patch

The complete change is below so that nobody has to redo it; I will open it as the PR as soon as a maintainer confirms the direction. 3 files, +39 -10.

<details><summary>git diff</summary>

```diff
diff --git a/test/distributions/test_distributions.py b/test/distributions/test_distributions.py
index d410ee7..deea36c 100644
--- a/test/distributions/test_distributions.py
+++ b/test/distributions/test_distributions.py
@@ -3820,6 +3820,22 @@ class TestDistributions(DistributionsTestCase):
                 f"Gumbel(loc={loc}, scale={scale})",
             )
 
+    def test_weibull_mode(self):
+        # the closed form only holds for concentration >= 1, below that a negative base
+        # raised to 1 / concentration gave nan or a value that is not the mode
+        scale = torch.tensor([2.0, 2.0, 2.0, 2.0])
+        concentration = torch.tensor([0.5, 1.0, 1.5, 3.0])
+        mode = Weibull(scale, concentration).mode
+        expected = torch.tensor(
+            [
+                nan,
+                0.0,
+                2.0 * (1.0 / 3.0) ** (2.0 / 3.0),
+                2.0 * (2.0 / 3.0) ** (1.0 / 3.0),
+            ]
+        )
+        self.assertEqual(mode, expected)
+
     def test_kumaraswamy_shape(self):
         concentration1 = torch.randn(2, 3).abs().requires_grad_()
         concentration0 = torch.randn(2, 3).abs().requires_grad_()
@@ -3843,6 +3859,17 @@ class TestDistributions(DistributionsTestCase):
 
     # Kumaraswamy distribution is not implemented in SciPy
     # Hence these tests are explicit
+    def test_kumaraswamy_mode(self):
+        # the log-space formula had the wrong signs and used concentration0 as the
+        # exponent, so the mode was nan for every parameter value
+        concentration1 = torch.tensor([2.0, 3.0, 1.0, 5.0, 0.5, 1.0])
+        concentration0 = torch.tensor([3.0, 2.0, 5.0, 1.0, 2.0, 1.0])
+        mode = Kumaraswamy(concentration1, concentration0).mode
+        expected = torch.tensor(
+            [(1.0 / 5.0) ** 0.5, (2.0 / 5.0) ** (1.0 / 3.0), 0.0, 1.0, nan, nan]
+        )
+        self.assertEqual(mode, expected)
+
     def test_kumaraswamy_mean_variance(self):
         c1_1 = torch.randn(2, 3).abs().requires_grad_()
         c0_1 = torch.randn(2, 3).abs().requires_grad_()
diff --git a/torch/distributions/kumaraswamy.py b/torch/distributions/kumaraswamy.py
index 9ad6620..d9a4609 100644
--- a/torch/distributions/kumaraswamy.py
+++ b/torch/distributions/kumaraswamy.py
@@ -81,11 +81,11 @@ class Kumaraswamy(TransformedDistribution):
 
     @property
     def mode(self) -> Tensor:
-        # Evaluate in log-space for numerical stability.
+        # Evaluate ((a - 1) / (a * b - 1)) ** (1 / a) in log-space for numerical stability.
         log_mode = (
-            self.concentration0.reciprocal() * (-self.concentration0).log1p()
-            - (-self.concentration0 * self.concentration1).log1p()
-        )
+            (self.concentration1 - 1).log()
+            - (self.concentration1 * self.concentration0 - 1).log()
+        ) / self.concentration1
         log_mode[(self.concentration0 < 1) | (self.concentration1 < 1)] = nan
         return log_mode.exp()
 
diff --git a/torch/distributions/weibull.py b/torch/distributions/weibull.py
index 22f48e2..29a57ec 100644
--- a/torch/distributions/weibull.py
+++ b/torch/distributions/weibull.py
@@ -1,7 +1,7 @@
 # mypy: allow-untyped-defs
 
 import torch
-from torch import Tensor
+from torch import nan, Tensor
 from torch.distributions import constraints
 from torch.distributions.exponential import Exponential
 from torch.distributions.gumbel import euler_constant
@@ -75,11 +75,13 @@ class Weibull(TransformedDistribution):
 
     @property
     def mode(self) -> Tensor:
-        return (
-            self.scale
-            * ((self.concentration - 1) / self.concentration)
-            ** self.concentration.reciprocal()
-        )
+        # the closed form only holds for concentration >= 1, below that the density is
+        # unbounded at zero, which lies outside the support, so the mode is undefined
+        concentration = self.concentration
+        defined = concentration >= 1
+        safe_conc = torch.where(defined, concentration, 2.0)
+        mode = self.scale * ((safe_conc - 1) / safe_conc) ** safe_conc.reciprocal()
+        return torch.where(defined, mode, nan)
 
     @property
     def variance(self) -> Tensor:
```

</details>

AI disclosure: I used an AI coding agent to help write this patch, the tests and this description. I have read the change and the tests myself and I will answer review comments personally.
